back/app/services/files/description_generator.rb

```diff
@@ -0,0 +1,199 @@
+  class DescriptionGenerator
+    delegate :generate_descriptions?, to: :class
+
+    # Generate and update the descriptions for the given file
+    # @return [Boolean] true if descriptions were updated, false if not.
+    def generate_descriptions!(file)
+      return false unless generate_descriptions?(file)
+
+      descriptions = generate_descriptions(file, tenant_locales)
+
+      file.with_lock do
+        # Only add descriptions if the file still does not have any descriptions.
+        # "still" because the descriptions could have been added while the descriptions
+        # were being generated.
+        next false if file.description_multiloc.present?
+
+        file.update!(description_multiloc: descriptions)
+      end
+    end
+
+    private
+
+    # Get configured locales for the application
+    #
+    # @return [Array<String>] Array of locale codes
+    def tenant_locales
+      @tenant_locales ||= AppConfiguration.instance.settings('core', 'locales')
+    end
+
+    # Generate descriptions using LLM for all configured locales
+    #
+    # @param content [String] The extracted file content
+    # @param locales [Array<String>] Array of locale codes
+    # @return [Hash] Hash with locale keys and description values
+    def generate_descriptions(file, locales)
+      schema = description_schema(locales)
+      chat = RubyLLM.chat(model: 'gpt-4.1').with_schema(schema)
+      prompt = build_prompt(file.name)
+
+      with_preprocessed_file_content(file) do |source|
+        chat.ask(prompt, with: source).content
+      end
+    end
+
+    # Yields the preprocessed file content to the block. If no preprocessing is needed,
+    # the original file content is passed. A block is used to ensure that any temporary
+    # files are properly closed and deleted after the block exits.
+    #
+    # @yield [source] The preprocessed file content
+    # @yieldparam [Tempfile, BaseFileUploader] source The preprocessed file content
+    def with_preprocessed_file_content(file, &)
+      preprocessed_file = preprocess_file(file)
+      source = preprocessed_file.is_a?(Tempfile) ? preprocessed_file.path : preprocessed_file.full_url
+      yield source
+    ensure
+      preprocessed_file.close! if preprocessed_file.is_a?(Tempfile)
+    end
+
+    # @return [Tempfile, BaseFileUploader]
+    def preprocess_file(file)
+      if file.text?
+        file.content
+      elsif file.image?
+        preprocess_image(file)
+      elsif previewable_as_pdf?(file)
+        preprocess_previewable_as_pdf(file)
+      else # rubocop:disable Lint/DuplicateBranch
+        file.content
+      end
+    end
+
+    # @return [Tempfile, BaseFileUploader]
+    def preprocess_previewable_as_pdf(file)
+      preview = file.preview
+
+      if preview.content.full_url.present?
+        preview.content
+      elsif preview.status == 'pending'
+        raise ApplicationJob::RetryInError.new('Preview is pending, try again later.', 30.seconds)
```

Comment:
I think this is okay, but 2 thoughts I want to share:
- Not 100% sure we want the retry logic to leak into the service. I guess it's fine because the interface is minimal and it's pretty convenient.
- If wanting to be true to the Ruby conventions (and some technical differences I honestly forgot), it's better to use `throw/catch` instead of `raise/rescue` for deliberate control flow, as opposed to error handling.